sources/sql_database/pg_replication/helpers.py

```diff
@@ -0,0 +1,528 @@
+            )
+            cur.consume_stream(consumer)
+        except StopReplication:
+            # all messages upto `upto_lsn` have been successfully consumed
+            consumed_all = True
+        finally:
+            cur.connection.close()
+            if consumed_all:
+                for i in consumer.data_items:
+                    yield i
+                self.last_commit_lsn = consumer.last_commit_lsn
+
+
+class MessageConsumer:
+    """Consumes messages from a ReplicationCursor.
+
+    Assumes all messages passed to __call__ have the same relation_id, i.e. they
+    belong to the same table.
+    """
+
+    def __init__(
+        self,
+        upto_lsn: int,
+        include_columns: Optional[Sequence[str]] = None,
+    ) -> None:
+        self.upto_lsn = upto_lsn
+        self.include_columns = include_columns
+
+        # data_items attribute maintains all data items
+        self.data_items: List[Union[TDataItem, DataItemWithMeta]] = []
+        # other attributes only maintain last-seen values
+        self.last_table_schema: TTableSchema
+        self.last_commit_ts: pendulum.DateTime
+        self.last_commit_lsn = None
+
+    def __call__(self, msg: ReplicationMessage) -> None:
+        """Processes message received from stream.
+
+        Breaks out of stream when `upto_lsn` is reached.
+        """
+        self.process_msg(msg)
+        if msg.data_start == self.upto_lsn:
+            raise StopReplication
+
+    def process_msg(self, msg: ReplicationMessage) -> None:
+        """Processes encoded replication message.
+
+        Identifies message type and decodes accordingly.
+        Message treatment is different for various message types.
+        """
+        op = (msg.payload[:1]).decode("utf-8")
+        if op == "B":
+            self.last_commit_ts = Begin(msg.payload).commit_ts
+        elif op == "C":
+            self.last_commit_lsn = msg.data_start
+        elif op == "R":
+            self.process_relation(Relation(msg.payload))
+        elif op == "I":
+            self.process_change(Insert(msg.payload), msg.data_start)
+        elif op == "U":
+            self.process_change(Update(msg.payload), msg.data_start)
+        elif op == "D":
+            self.process_change(Delete(msg.payload), msg.data_start)
+        elif op == "T":
+            logger.warning(
+                "The truncate operation is currently not supported. "
+                "Truncate replication messages are ignored."
+            )
+
+    def process_relation(self, decoded_msg: Relation) -> None:
+        """Processes a replication message of type Relation.
+
+        Stores table schema information from Relation message in object state
+        and adds meta data item to the relation's list of data items to update
+        the table schema.
+        """
+        # store table schema information
+        columns = {c.name: _to_dlt_column_schema(c) for c in decoded_msg.columns}
```

Comment:
when do we get this message? is it coming with every change? it looked like it in jsonpgout

Reply:
See my previous [comment](https://github.com/dlt-hub/verified-sources/pull/392#discussion_r1514523102).